### DISCUSSION TITLE
route cause pydantic.errors.PydanticUndefinedAnnotation

### DISCUSSION THREAD

**Discussion by PaleNeutron** [fastapi/fastapi#11423] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Description

Base on structure documented here: https://sqlmodel.tiangolo.com/tutorial/code-structure/

And create output models for fastapi as documented: https://sqlmodel.tiangolo.com/tutorial/fastapi/relationships/

Just add routes.py and import it in app.py will raise 

```
pydantic.errors.PydanticUndefinedAnnotation: name 'Hero' is not defined
```

### Example Code

<details>
<summary>Code from document</summary>

hero_model.py

```python
from typing import TYPE_CHECKING, Optional

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from .team_model import Team


class HeroBase(SQLModel):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(index=True)
    secret_name: str
    age: Optional[int] = Field(default=None, index=True)

    team_id: Optional[int] = Field(default=None, foreign_key="team.id")
    
class Hero(HeroBase, table=True):
    team: Optional["Team"] = Relationship(back_populates="heroes")


class HeroRead(HeroBase):
    teams: list["Team"]
```

team_model.py

```python
from typing import TYPE_CHECKING, List, Optional

from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from .hero_model import Hero


class TeamBase(SQLModel):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(index=True)
    headquarters: str

class Team(TeamBase, table=True):
    heroes: List["Hero"] = Relationship(back_populates="team")

class TeamWithHeros(TeamBase):
    heroes: List["Hero"]
```

</details>

routes.py

```python
from fastapi import APIRouter

from .team_model import Team, TeamWithHeros

router = APIRouter()


@router.get("/teams")
async def read_teams()->"TeamWithHeros":   # <---- If we change `TeamWithHeros` to `Team`, it works
    return {"teams": "teams"}
```

app.py

```python
from .hero_model import Hero
from .team_model import Team


from .routes import router  # THIS LINE CAUSE ERROR

def create_heroes():
        team_z_force = Team(name="Z-Force", headquarters="Sister Margaret's Bar")

def main():
    create_heroes()


if __name__ == "__main__":
    main()
```

### Operating System

Linux

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

Python 3.10.12 / Python 3.12.3

**Comment by YuriiMotov** [comment, marked as answer]
I've met this problem before. As I remember, it comes from Pydantic (I mean you will have the same problem if you use Pydantic models the same way). And I didn't find any ways to solve it except putting both models to one module

**Reply by PaleNeutron** [reply]
Yes, I found one module is the only solution too.

But I don't think it is a bug with Pydantic, since if I remove the following line, the code works.

```
from .routes import router  # THIS LINE CAUSE ERROR
```

or even more simply, replace `TeamWithHeros` with `Team` in route typehint, the example works too.

I think there must be something not match the design of pydantic in router registration.

**Reply by YuriiMotov** [reply]
That works without `from .routes import router` because you don't use `TeamWithHeros` anywhere.
Just try using it and you will see that it doesn't work:

main.py
```
from hero_model import Hero, HeroRead
# from routes import router  # THIS LINE CAUSE ERROR
from team_model import Team, TeamWithHeros

def main():
    team_z_force = Team(name="Z-Force", headquarters="Sister Margaret's Bar")
    hero = Hero(name="hero 1", secret_name="h1", team=team_z_force)
    team_with_heroes = TeamWithHeros.model_validate(team_z_force, from_attributes=True)

if __name__ == "__main__":
    main()
```
